Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+  if (mPrint) {
+    std::cout << "orbit_c: " << orbit_c << "  orbit_p: " << orbit_p << std::endl;
+  }
+  if ((orbit_c >= orbit_p) && ((orbit_c - orbit_p) > 1)) {
+    return;
+  }
+
+  if (mPrint) {
+    std::cout << "Merging digits from " << currentBufId << " (orbit=" << currentBuffer.orbit << ") into "
+              << previousBufId << " (orbit=" << previousBuffer.orbit << ")\n";
+  }
+
+  for (size_t i = 0; i < currentBuffer.digits.size(); i++) {
+    MergerDigit& d1 = currentBuffer.digits[i];
+
+    // skip digits that do not start at the beginning of the time window
+    Digit::Time startTime = d1.digit.getTime();
+    if (startTime.sampaTime != 0) {
+      continue;
+    }
+
+    for (size_t j = 0; j < previousBuffer.digits.size(); j++) {
+      MergerDigit& d2 = previousBuffer.digits[j];
+      if (checkDigits(d1, d2)) {
+        break;
+      }
+    }
+  }
+}
+
+void Merger::setOrbit(int feeId, uint32_t orbit, bool stop)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].setOrbit(orbit, stop);
+}
+
+void Merger::setDigitHandler(std::function<void(const Digit&)> h)
+{
+  for (int feeId = 0; feeId <= MCH_MERGER_FEEID_MAX; feeId++) {
+    mergers[feeId].setDigitHandler(h);
+  }
+}
+
+void Merger::addDigit(int feeId, int solarId, int dsAddr, int chAddr,
+                      int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].getCurrentBuffer().digits.emplace_back(MergerDigit{o2::mch::Digit(deId, padId, adc, time),
+                                                                    stopTime, false, solarId, dsAddr, chAddr});
+}
+
+void Merger::mergeDigits(int feeId)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].mergeDigits();
+}
+
+//=======================
+// Data decoder
+void DataDecoder::decodeBuffer(gsl::span<const std::byte> page)
+{
+  size_t ndigits{0};
+
+  uint8_t isStopRDH = 0;
+  uint32_t orbit;
+  uint32_t feeId;
+  uint32_t linkId;
+
+  const auto storeDigit = [&](const Digit& d) {
+    mOutputDigits.emplace_back(d);
+    if (mPrint) {
+      std::cout << "[storeDigit]: digit stored, mOutputDigits size: " << mOutputDigits.size() << std::endl;
+    }
+  };
+
+  auto channelHandler = [&](DsElecId dsElecId, uint8_t channel, o2::mch::raw::SampaCluster sc) {
+    if (mChannelHandler) {
+      mChannelHandler(dsElecId, channel, sc);
+    }
+
+    if (mDs2manu) {
+      channel = ds2manu(int(channel));
+    }
+
+    double digitadc(0);
+    for (auto& s : sc.samples) {
+      digitadc += s;
+    }
+
+    int deId{-1};
+    int dsIddet{-1};
+    if (auto opt = mElec2Det(dsElecId); opt.has_value()) {
+      DsDetId dsDetId = opt.value();
+      dsIddet = dsDetId.dsId();
+      deId = dsDetId.deId();
+    }
+
+    int padId = -1;
+    const Segmentation& segment = segmentation(deId);
+    if ((&segment) == nullptr) {
```

Comment:
I believe the call to segmentation(invalid_deid) will simply crash, so no need for a check. Or said otherwise, the check of deId validity should be made beforehand

Reply:
Is there an existing function for checking the validity of a DetID? Or could this be added to the MCH mapping library? This is a recurring problem in different places, like the QC...

Reply:
Not quite yet, but there's a deIdsForAllMCH array defined in MCHRaw/Mapper.h that could easily be used to define such a function. I'll have a look.

Reply:
Ok, then I'll use the deIdsForAllMCH array for the beginning, and we can introduce the function when available.